sdk/tests/conformance2/reading/read-pixels-upon-pixel-pack-buffer.html

```diff
@@ -0,0 +1,219 @@
+    gl.readPixels(0, 0, 1, 1, gl.RGBA, gl.UNSIGNED_BYTE, array);
+    wtu.glErrorShouldBe(gl, gl.NO_ERROR);
+    gl.readPixels(0, 0, 1, 1, gl.RGBA, gl.UNSIGNED_BYTE, 0);
+    wtu.glErrorShouldBe(gl, gl.INVALID_OPERATION, "no pixel pack buffer bound");
+
+    gl.deleteBuffer(buffer);
+}
+
+function calculatePaddingBytes(bytesPerPixel, packAlignment, width)
+{
+    var padding = 0;
+    switch (packAlignment) {
+    case 1:
+    case 2:
+    case 4:
+    case 8:
+        padding = (bytesPerPixel * width) % packAlignment;
+        if (padding > 0)
+            padding = packAlignment - padding;
+        break;
+    default:
+        return -1;
+    }
+    return padding;
+}
+
+function runTestIteration(format, type, packAlignment, width, height)
+{
+    debug("Testing PACK_ALIGNMENT = " + packAlignment + ", width = " + width + ", height = " + height);
+    gl.clearColor(1, 0.4, 0, 1);
+    gl.clear(gl.COLOR_BUFFER_BIT);
+    gl.pixelStorei(gl.PACK_ALIGNMENT, packAlignment);
+    wtu.glErrorShouldBe(gl, gl.NO_ERROR);
+    var bytesPerPixel = 4;
+    var padding = calculatePaddingBytes(bytesPerPixel, packAlignment, width);
+    var size = bytesPerPixel * width * height + padding * (height - 1);
+    if (size < 0)
+        size = 0;
+    var buffer = gl.createBuffer();
+    gl.bindBuffer(gl.PIXEL_PACK_BUFFER, buffer);
+    gl.bufferData(gl.PIXEL_PACK_BUFFER, size, gl.STATIC_DRAW);
+    var offset = 0;
+    gl.readPixels(0, 0, width, height, format, type, offset);
+    if (width < 0 || height < 0) {
+        wtu.glErrorShouldBe(gl, gl.INVALID_VALUE);
+        return;
+    }
+    wtu.glErrorShouldBe(gl, gl.NO_ERROR);
+
+    if (!size)
+        return;
+    var arrBuffer = new ArrayBuffer(size);
+    gl.getBufferSubData(gl.PIXEL_PACK_BUFFER, 0, arrBuffer);
+    var array = new Uint8Array(arrBuffer);
+
+    // Check the last pixel of the last row.
+    var bytesPerRow = width * bytesPerPixel + padding;
+    var pos = bytesPerRow * (height - 1) + (width - 1) * bytesPerPixel;
+    var numComponents = bytesPerPixel;
+    for (var i = 0; i < numComponents; ++i)
+        pixel[i] = array[pos + i];
+    for (var i = numComponents; i < 4; ++i)
+        pixel[i] = 0;
+    expectedColor = [255, 102, 0, 255];
+    shouldBeNonNull("expectedColor");
+    shouldBe("pixel", "expectedColor");
+}
+
+function testPackAlignment()
+{
+    debug("");
+    debug("Verify that reading pixels to PIXEL_PACK buffer works fine with various PACK_ALIGNMENT values.");
+    shouldBeNonNull("program = wtu.setupProgram(gl, ['vshader', 'fshader'], ['pos', 'colorIn'])");
+
+    var format = gl.RGBA;
+    var type = gl.UNSIGNED_BYTE;
+
+    debug("format: RGBA, type: UNSIGNED_BYTE");
+    runTestIteration(format, type, 1, 1, 2);
+    runTestIteration(format, type, 8, 1, 2);
+    runTestIteration(format, type, 1, 0, 0);
+    runTestIteration(format, type, 1, 0, -1);
+    runTestIteration(format, type, 1, -1, 0);
+    runTestIteration(format, type, 1, -1, -1);
```

Comment:
You already tested negative height/width in validatePixelPackBufferAndParameters(). More interesting test cases would be various pack alignment, since this function name is testPackAlignment() anyway.